### PR TITLE
Calendar improvements

### LINKED ISSUE #21713: Calendar: Window marked as "modified" after adding event, no prompt on close
To reproduce:

1. Open Calendar
2. Create a new event
3. Notice that the Calendar window now has the "x" close button with "..."
4. Click the close button
5. Notice how it closes without prompting

The window modified state is supposed to indicate that there's some unsaved state. If I close the window with unsaved state, I expect to be prompted about saving it :)

### PR DESCRIPTION
**Pull request by david072**
This PR contains some small improvements for the calendar application:
- Added a date picker dialog to make selecting dates easier
- Refactored the code a little bit to make calendar events easier to work with
- Removed some unused code
- Ask about unsaved changes when attempting to close the window (fixes #21713)

### CODE REVIEW

File: Userland/Libraries/LibGUI/CMakeLists.txt [hunk 1]
```diff
@@ -1,4 +1,5 @@
 stringify_gml(AboutDialog.gml AboutDialogGML.h about_dialog_gml)
+stringify_gml(DatePickerDialog.gml  DatePickerDialogGML.h date_picker_dialog_gml)
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibGUI/CMakeLists.txt:2]: Nit-pick: There's an extra space here

File: Userland/Libraries/LibGUI/CMakeLists.txt [hunk 2]
```diff
@@ -143,6 +145,7 @@ set(GENERATED_SOURCES
     FontPickerDialogGML.h
     IncrementalSearchBannerGML.h
     PasswordInputDialogGML.h
+    DatePickerDialogGML.h
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibGUI/CMakeLists.txt:148]: Please keep this list alphabetised

File: Userland/Libraries/LibGUI/DatePickerDialog.h [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+            __Count,
+        };
+
+        static NonnullRefPtr<MonthListModel> create() { return adopt_ref(*new MonthListModel); }
+        virtual ~MonthListModel() override = default;
+
+        virtual int row_count(const GUI::ModelIndex& = GUI::ModelIndex()) const override { return 12; }
+        virtual int column_count(const GUI::ModelIndex& = GUI::ModelIndex()) const override { return Column::__Count; }
+        virtual ErrorOr<String> column_name(int) const override;
+        virtual GUI::Variant data(const GUI::ModelIndex&, GUI::ModelRole) const override;
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibGUI/DatePickerDialog.h]: Nit-pick: Please use east const: `Foo const&` not `const Foo&`.

(clang-format is not great at this, and the IDE I use also messes this up constantly, so it's easily done.)

File: Userland/Libraries/LibGUI/DatePickerDialog.h [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+#include <AK/String.h>
+#include <LibCore/DateTime.h>
+#include <LibGUI/Calendar.h>
+#include <LibGUI/Dialog.h>
+#include <LibGUI/Model.h>
+
+namespace GUI {
+
+class DatePickerDialog : public Dialog {
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibGUI/DatePickerDialog.h:17]: Not a comment on this line actually, but please add `DatePickerDialog` to `LibGUI/Forward.h`.

File: Userland/Libraries/LibCore/DateTime.cpp [hunk 5]
```diff
@@ -85,6 +85,16 @@ void DateTime::set_time(int year, int month, int day, int hour, int minute, int
...
+void DateTime::set_time_only(int hour, int minute, int second)
+{
+    set_time(year(), month(), day(), hour, minute, (second == -1) ? this->second() : second);
+}
+
+void DateTime::set_date(Core::DateTime other)
+{
+    set_time(other.year(), other.month(), other.day(), hour(), minute(), second());
+}
+
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibCore/DateTime.cpp:97]: A couple of points on this:
- These DateTime changes are distinct, so they belong in a separate commit before this one.
- Please use `Optional<int>` instead of giving `-1` a special meaning. (You can use `{}` as the default value to mean an empty Optional, and then something like `second.value_or(this->second())` to get the value out.)
- `DateTime::set_date()` should take a const reference to `other` instead of copying it.

File: Userland/Applications/Calendar/AddEventDialog.cpp [hunk 6]
```diff
@@ -53,8 +53,8 @@ AddEventDialog::AddEventDialog(Core::DateTime date_time, EventManager& event_man
     auto widget = set_main_widget<GUI::Widget>();
     widget->load_from_gml(add_event_dialog_gml).release_value_but_fixme_should_propagate_errors();
 
-    auto event_title_textbox = widget->find_descendant_of_type_named<GUI::TextBox>("event_title_textbox");
-    event_title_textbox->set_focus(true);
+    auto& event_title_textbox = *widget->find_descendant_of_type_named<GUI::TextBox>("event_title_textbox");
+    event_title_textbox.set_focus(true);
```
Comment (AtkinsSJ) [inline Userland/Applications/Calendar/AddEventDialog.cpp:57]: These convert-pointer-to-reference changes aren't really related to this commit, and would be better separate.

File: Userland/Applications/Calendar/AddEventDialog.gml [hunk 7]
```diff
@@ -123,6 +111,12 @@
 
         @GUI::Layout::Spacer {}
 
+        @GUI::Button {
+            name: "cancel_button"
+            text: "Cancel"
+            fixed_size: [80, 20]
+        }
```
Comment (AtkinsSJ) [inline Userland/Applications/Calendar/AddEventDialog.gml:118]: Adding the cancel button also wants to be a separate commit.

File: Userland/Applications/Calendar/AddEventDialog.gml [hunk 8]
```diff
@@ -51,21 +51,15 @@
...
             min: 1
-            max: 12
+            max: 24
         }
 
         @GUI::SpinBox {
             name: "start_minute"
             fixed_size: [40, 20]
-            min: 1
+            min: 0
```
Comment (AtkinsSJ) [inline Userland/Applications/Calendar/AddEventDialog.gml:60]: Fixing this and the `end_minute` should be a separate commit.

File: Userland/Applications/Calendar/EventManager.cpp [hunk 9]
```diff
@@ -41,21 +40,65 @@ ErrorOr<void> EventManager::save(FileSystemAccessClient::File& file)
...
 
     return {};
 }
 
+ErrorOr<JsonArray> EventManager::serialize_events()
+{
+    JsonArray result;
+    for (auto const& event : m_events) {
+        JsonObject object;
+        object.set("start", JsonValue(event.start.to_deprecated_string("%Y-%m-%dT%H:%M:%S"sv)));
```
Comment (AtkinsSJ) [inline Userland/Applications/Calendar/EventManager.cpp:57]: This date format gets used in several places in this file, can we make it a constant instead?

File: Userland/Libraries/LibGUI/DatePickerDialog.gml [hunk 10]
```diff
@@ -7,7 +7,7 @@
         fixed_height: 20
         fill_with_background_color: true
         layout: @GUI::HorizontalBoxLayout {}
-        
+
```
Comment (AtkinsSJ) [inline Userland/Libraries/LibGUI/DatePickerDialog.gml:10]: Please make these corrections in the commit the DatePickerDialog was added instead.

File: Userland/Applications/Calendar/EventManager.cpp [hunk 11]
```diff
@@ -24,12 +24,14 @@ OwnPtr<EventManager> EventManager::create()
 void EventManager::add_event(Event event)
 {
     m_events.append(move(event));
+    m_dirty = true;
```
Comment (AtkinsSJ) [inline Userland/Applications/Calendar/EventManager.cpp:27]: You removed this dirty state a couple of commits ago. Please drop that earlier commit, instead of removing it and then adding it again later.

**Review by AtkinsSJ — CHANGES_REQUESTED** [review]
Sorry this has sat around for a while unnoticed! There are some really nice changes in here. In future, I'd really recommend adding screenshots to a PR if you can. (You can just drag&drop or paste an image into the text box on GitHub.) It really helps get people's attention. That date picker dialog looks great, so show it off! :^)

I've ended up repeating myself a bit. One theme is that we like to have "atomic" commits. That is, each commit is a single change. Adding the DatePickerDialog, fixing a bug, reformatting or moving some code: each of these is a separate change. A few places you've combined multiple changes into one commit, which is easily done when linters are nagging at you about little things. But the reason we avoid that is it both makes it easier to identify where a bug was introduced, and makes it easier to revert a single change if needed. It definitely takes some getting used to, to work like this, but it's nice. :^)

I've no idea how familiar you are with modifying previous commits, but [this video](https://www.youtube.com/watch?v=ElRzTuYln0M) is helpful. Failing that, ask on Discord. :^)

**Comment by david072** [comment]
Okay, thanks for the review!

> Sorry this has sat around for a while unnoticed!

I mean I didn't advertise it either so... :shrug:

Sorry about all those small things, but I think I got all of them.
I had to update the branch with master though since it wouldn't build otherwise (see #build-discussion on Discord), but as you can see it didn't work quite as expected... Can I fix this or do I have to just create a new branch and cherry-pick the commits over?

**Comment by david072** [comment]
Sorry to everyone who got a review request. Please ignore :slightly_smiling_face:

**Comment by david072** [comment]
On second look, I think this is completely fucked... I'll create a new PR and try to transfer the changes.
